### DISCUSSION TITLE
Getting a QR code with circles and logo?

### DISCUSSION THREAD

**Discussion by achleitner** [chillerlan/php-qrcode#178] [General]
Hello, so unfortunately I cant get any of the examples provided to work. Mostly getting server erros when copying and pasting.
I managed to output this simple qrcode using below code. But even that isnt working correctly as it isnt outputtig circles as it is set in the options. I would like to eventually get a qrcode with circles and a logo in the middle but it seems like quite the challenge as I have no idea why all the examples are failing, when the library is apparently loaded correctly. So where to start getting this going? Any help appreciated. Thanks

```php
$options = new QROptions([
	'scale'               => 7,
	'imageBase64'         => true,
	'imageTransparent'    => true,
	'drawCircularModules' => true,
	'circleRadius'        => 0.6,	
]);

$qrcode = new QRCode($options);
$qrcode = $qrcode->render($final_url);
<img src="<?php echo $qrcode;?>" />
```

**Comment by codemasher** [comment]
Hmm, are you trying to run a release version (e.g. 4.3.x) with one of the examples from dev-main? that would explain the server errors you're getting and also why it won't work.

**Reply by achleitner** [reply]
Hello thanks, I installed it using composer and used the examples from https://github.com/chillerlan/php-qrcode/tree/main/examples Mostly just getting nothing returned. For example this code:

```php
$options = new QROptions([
	'version' => 4,
	'eccLevel'   => QRCode::ECC_M,
	'addLogoSpace'        => true,
	'logoSpaceWidth'      => 8,
	'logoSpaceHeight'     => 8,
	'scale'               => 6,
	'imageBase64'         => true,
	'imageTransparent'    => true,
	'drawCircularModules' => true,
	'circleRadius'        => 0.4,
]);
$qrcode = new QRCode($options);
$qrcode = $qrcode->render($final_url);

echo '<div><img src="'.$qrcode.'" /></div>';
```


outputs this qr code:

![Download](https://user-images.githubusercontent.com/42497073/214866560-a55dca4f-9edf-4ffc-8e52-fc9f536daa15.png)

So I am kinda lost since this seems so basic but there is no logo space or circular modules. I am probably missing something basic :) Thanks

**Reply by codemasher** [reply]
I'm wondering if you have `error_reporting(0)` or something set in your script because with the given options you should get an exception first and foremost: `ECC level "H" required to add logo space` - with that fixed, everything seems fine to me:

```php
$options = new QROptions([
	'version'             => 4,
	'eccLevel'            => EccLevel::H, // ECC level H is required for logo space
	'addLogoSpace'        => true,
	'logoSpaceWidth'      => 8,
	'logoSpaceHeight'     => 8,
	'scale'               => 6,
	'imageBase64'         => true,
	'imageTransparent'    => true,
	'drawCircularModules' => true,
	'circleRadius'        => 0.4,
]);

$qrcode = new QRCode($options);
$qrcode = $qrcode->render($final_url);

echo '<div><img style="width: 300px; height: 300px;" alt="qrcode" src="'.$qrcode.'" /></div>';
```

**Comment by achleitner** [comment]
Ok thanks, this isnt returning anything for me unfortunately. Seems like I am using the wrong version for the examples as you said or something. Are there any detailed instructions on how to use 4.3?

I can see in 4.3 for example "'eccLevel'   => QRCode::ECC_L," in an example, but you are using "'eccLevel'   => EccLevel::H, " so how can I find out what to use? Thanks

**Reply by achleitner** [reply]
I tried  "composer require chillerlan/php-qrcode:dev-main#fbb4811094940f125657297b74d34b541512c501" and seeing some success now. Haha sorry for bothering but this had me a bit confused.


![Download](https://user-images.githubusercontent.com/42497073/214886899-8faad2e3-da8c-426b-957f-3b3e8e2136b8.png)

**Reply by codemasher** [reply]
Several of the `QRCode::*` constants have been deprecated and been moved to their respective submodules for the upcoming v5, this is one reason why the v4 release versions and dev-examples don't mix, among other things.
